flow/raster_cache.h

```diff
@@ -42,6 +53,44 @@ class RasterCacheResult {
 
 struct PrerollContext;
 
+// The |RasterCacheImageDelegate| creates and retruns a cached result of
+// rendering the supplied |Layer| or |SkPicture|.
+class RasterCacheImageDelegate {
+ public:
+  virtual std::unique_ptr<RasterCacheResult> RasterizePicture(
+      SkPicture* picture,
+      GrContext* context,
+      const SkMatrix& ctm,
+      SkColorSpace* dst_color_space,
+      bool checkerboard) const = 0;
+
+  virtual std::unique_ptr<RasterCacheResult> RasterizeLayer(
+      PrerollContext* context,
+      Layer* layer,
+      const SkMatrix& ctm,
+      bool checkerboard) const = 0;
+};
+
+// The |SkRasterCacheImageDelegate| creates and returns rendered results
+// for |Layer| and |SkPicture| objects stored as |SkImage| objects.
+class SkRasterCacheImageDelegate : public RasterCacheImageDelegate {
+ public:
+  std::unique_ptr<RasterCacheResult> RasterizePicture(
+      SkPicture* picture,
+      GrContext* context,
+      const SkMatrix& ctm,
+      SkColorSpace* dst_color_space,
+      bool checkerboard) const override;
+
+  std::unique_ptr<RasterCacheResult> RasterizeLayer(
```

Comment:
Why not make `RasterizePicture` and `RasterizeLayer` virtual functions of `RasterCache` class so you can have `MockRasterCache` instead of `MockRasterCacheImageDelegate`? It seems that we can save a few classes in that way.

Reply:
In general I prefer pure virtual methods of providing alternate implementations, but this change does make the overall code diffs much simpler so I'll go with the quick and easy.

Refactoring done.